### PR TITLE
Accumulate all features from cargo tree output

### PR DESCRIPTION
**Pull request by gferon**
Consider the following `cargo tree --target aarch64-apple-darwin --workspace --prefix=none --format="{p} {f}"` output (which you can obtain with a dependency to `matrix-sdk = 0.6` for example):

```
js_int v0.2.2 default,std
js_int v0.2.2 default,serde,std
js_int v0.2.2 default,serde,std (*)
js_int v0.2.2 default,serde,std (*)
js_int v0.2.2 default,serde,std (*)
```

without this patch, the `crate_features` map will only contain `default` and `std` which means the dependents of the `js_int` crate will fail to compile when they require serialization. 

Since universe doesn't really support the compilation of the same crate with different feature sets (yet), it makes more sense to me to pick the biggest set of features.

### CODE REVIEW

**Comment by gferon** [comment]
I suppose this is somewhat related to https://github.com/bazelbuild/rules_rust/issues/1609

**Comment by gferon** [comment]
Note: Looks like the CI failure is unrelated to this change

**Comment by gferon** [comment]
Confirmed to fix #1850

**Comment by gferon** [comment]
Absolutely, will do!

**Comment by UebelAndre** [comment]
@wmatthews-google would you also be able to take a look?

**Comment by wmatthews-google** [comment]
This basically looks good to me. Thanks!

+1 for add a unit test (or perhaps extend `parse_features_from_cargo_tree_output_prefix_none` in crate_universe/src/metadata.rs.)

I think we could probably take the repro example from #1850 and make it an integration test case (crate_universe/tests/cargo_integration_test.rs), but I can do this as a followup PR if you'd like.

**Comment by UebelAndre** [comment]
Looks like there are some failing CI jobs though
https://buildkite.com/bazel/rules-rust-rustlang/builds/8301#0186ffc5-40d5-4f54-82b1-334eab577698

**Comment by gferon** [comment]
@UebelAndre is the cargo package cache maybe pinned somewhere in CI?

File: crate_universe/tests/cargo_integration_test.rs [hunk 1]
```diff
@@ -261,3 +261,36 @@ fn feature_generator_workspace() {
...
+    // we just skip this test case.
+    use std::net::ToSocketAddrs;
+    if "github.com:443".to_socket_addrs().is_err() {
+        eprintln!("This test case requires network access. Skipping!");
+        return;
+    }
+
+    let runfiles = runfiles::Runfiles::create().unwrap();
+    let metadata = run(
+        "crate_combined_features(",
```
Comment (UebelAndre) [inline crate_universe/tests/cargo_integration_test.rs:279]:
```suggestion
        "crate_combined_features",
```

File: crate_universe/tests/cargo_integration_test.rs [hunk 2]
```diff
@@ -261,3 +261,36 @@ fn feature_generator_workspace() {
...
+    // other environments) disallow or don't support this. In those cases,
+    // we just skip this test case.
+    use std::net::ToSocketAddrs;
+    if "github.com:443".to_socket_addrs().is_err() {
+        eprintln!("This test case requires network access. Skipping!");
+        return;
+    }
+
+    let runfiles = runfiles::Runfiles::create().unwrap();
+    let metadata = run(
```
Comment (UebelAndre) [inline crate_universe/tests/cargo_integration_test.rs:278]: Perhaps `run` should run `cargo fetch` before calling `splice` to make sure the registry is up to date? I bet `cargo tree --locked` does not update the index? which might be a bug there?

Maybe the long term fix is to have https://github.com/bazelbuild/rules_rust/blob/0.19.0/crate_universe/src/splicing.rs#L419-L451 call `cargo fetch` or something that updates the index in the event that the function does not actually call `cargo generate-lockfile`?

Comment (gferon) [reply crate_universe/tests/cargo_integration_test.rs:278]: I did a quick test and `cargo tree --locked` does update the registry properly (at least on my machine) even after clearing `~/.cargo/registry`.

Comment (gferon) [reply crate_universe/tests/cargo_integration_test.rs:278]: One more feature I didn't really think of... [`matrix-sdk = 0.6.2` requires Rust 1.65](https://github.com/matrix-org/matrix-rust-sdk/blob/main/Cargo.toml#L20), thanks for figuring this out.

**Comment by gferon** [comment]
Changed the test to use another example, which duplicate `serde` with different feature sets in the output of `cargo tree`. This works with both Rust 1.59 and 1.68. I think it's good to go! 🚀

**Pull request merged**